
In order to better understand our data, we start by visualising and exploring our data which are the first step in data anlysis. 


# Data

In [1]:
import pandas as pd 
import warnings
warnings.filterwarnings("ignore")

In [2]:
DATA = pd.read_csv('train.csv')
DATA

ImgId                                              title  \
0      B000HYL1V6                   TUNGSTEN SOLDER PICK WITH HANDLE   
1      B00006HXWY  Write Right 98167 Screen Protector for Sony T615C   
2      B000GAWSBS  Casio Mens DBC310-1 Databank 300 Digital Watch...   
3      B000040JOL  Factory-Reconditioned DEWALT DW260KR Heavy-Dut...   
4      B00006IB78                             Energizer 2 in 1 Light   
...           ...                                                ...   
46224  B0000AV1J7  Berkley Trilene XL 1000 - yard Fishing Line, C...   
46225  B0006MZPLG       Badger Air-Brush Company Basic Spray Gun Set   
46226  B000GBO8SS  Fimo Soft Polymer Clay 2 Ounces-8020-53 Tropic...   
46227  B0000VMA16               Wilton Icing Colors 1 Oz: Rose Petal   
46228  B000H6UOSK  Polyform Sculpey III Polymer Clay 2 Oz: French...   

                                             description  \
0      Solder Pick for picking up molten solder when ...   
1      We all screen. And we all need to protect thos...   
2      Bringing you precision at a glance, the Casio ...   
3      Factory-Reconditioned DEWALT DW260KR Heavy-Dut...   
4      This twoway light features a bright flashlight...   
...                                                  ...   
46224  Smooth - casting Berkley Trilene XL Fishing Li...   
46225  An easy to use, top quality spray gun for the ...   
46226  Fimo Soft Polymer Clay is easier to use than b...   
46227  Icing Colors. This plastic bottle contains 1 o...   
46228  Sculpey 3 Polymer Clay is America's original o...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
46224          Sports & Outdoors  
46225      Arts, Crafts & Sewing  
46226      Arts, Crafts & Sewing  
46227     Grocery & Gourmet Food  
46228      Arts, Crafts & Sewing  

[46229 rows x 4 columns]

### Construct dataframe that contain just the features that we need(product name, product description and category that belongs to)

In [3]:
#Create dataframe
column_names = ['name','description','categories']
df = pd.DataFrame(columns = column_names)

In [4]:
df

Empty DataFrame
Columns: [name, description, categories]
Index: []

In [5]:
df.name = DATA.title
df.description = DATA.description 
df.categories = DATA.categories

In [6]:
df

name  \
0                       TUNGSTEN SOLDER PICK WITH HANDLE   
1      Write Right 98167 Screen Protector for Sony T615C   
2      Casio Mens DBC310-1 Databank 300 Digital Watch...   
3      Factory-Reconditioned DEWALT DW260KR Heavy-Dut...   
4                                 Energizer 2 in 1 Light   
...                                                  ...   
46224  Berkley Trilene XL 1000 - yard Fishing Line, C...   
46225       Badger Air-Brush Company Basic Spray Gun Set   
46226  Fimo Soft Polymer Clay 2 Ounces-8020-53 Tropic...   
46227               Wilton Icing Colors 1 Oz: Rose Petal   
46228  Polyform Sculpey III Polymer Clay 2 Oz: French...   

                                             description  \
0      Solder Pick for picking up molten solder when ...   
1      We all screen. And we all need to protect thos...   
2      Bringing you precision at a glance, the Casio ...   
3      Factory-Reconditioned DEWALT DW260KR Heavy-Dut...   
4      This twoway light features a bright flashlight...   
...                                                  ...   
46224  Smooth - casting Berkley Trilene XL Fishing Li...   
46225  An easy to use, top quality spray gun for the ...   
46226  Fimo Soft Polymer Clay is easier to use than b...   
46227  Icing Colors. This plastic bottle contains 1 o...   
46228  Sculpey 3 Polymer Clay is America's original o...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
46224          Sports & Outdoors  
46225      Arts, Crafts & Sewing  
46226      Arts, Crafts & Sewing  
46227     Grocery & Gourmet Food  
46228      Arts, Crafts & Sewing  

[46229 rows x 3 columns]

### Save the new dataframe as .csv file

In [7]:
df.to_csv(r'products_data.csv', index=False)

In [8]:
# read csv file
products_data=pd.read_csv('products_data.csv')

In [9]:
products_data

name  \
0                       TUNGSTEN SOLDER PICK WITH HANDLE   
1      Write Right 98167 Screen Protector for Sony T615C   
2      Casio Mens DBC310-1 Databank 300 Digital Watch...   
3      Factory-Reconditioned DEWALT DW260KR Heavy-Dut...   
4                                 Energizer 2 in 1 Light   
...                                                  ...   
46224  Berkley Trilene XL 1000 - yard Fishing Line, C...   
46225       Badger Air-Brush Company Basic Spray Gun Set   
46226  Fimo Soft Polymer Clay 2 Ounces-8020-53 Tropic...   
46227               Wilton Icing Colors 1 Oz: Rose Petal   
46228  Polyform Sculpey III Polymer Clay 2 Oz: French...   

                                             description  \
0      Solder Pick for picking up molten solder when ...   
1      We all screen. And we all need to protect thos...   
2      Bringing you precision at a glance, the Casio ...   
3      Factory-Reconditioned DEWALT DW260KR Heavy-Dut...   
4      This twoway light features a bright flashlight...   
...                                                  ...   
46224  Smooth - casting Berkley Trilene XL Fishing Li...   
46225  An easy to use, top quality spray gun for the ...   
46226  Fimo Soft Polymer Clay is easier to use than b...   
46227  Icing Colors. This plastic bottle contains 1 o...   
46228  Sculpey 3 Polymer Clay is America's original o...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
46224          Sports & Outdoors  
46225      Arts, Crafts & Sewing  
46226      Arts, Crafts & Sewing  
46227     Grocery & Gourmet Food  
46228      Arts, Crafts & Sewing  

[46229 rows x 3 columns]

In [10]:
products_data.shape

(46229, 3)

#### Our data contain 46229 products and  3 features 

In [11]:
#Number of labels
cat = pd.DataFrame(products_data['categories'].unique())
cat.shape

(21, 1)

In [12]:
cat

0
0       Arts, Crafts & Sewing
1   Cell Phones & Accessories
2   Clothing, Shoes & Jewelry
3    Tools & Home Improvement
4      Health & Personal Care
5               Baby Products
6                        Baby
7        Patio, Lawn & Garden
8                      Beauty
9           Sports & Outdoors
10                Electronics
11            All Electronics
12                 Automotive
13               Toys & Games
14                 All Beauty
15            Office Products
16                 Appliances
17        Musical Instruments
18    Industrial & Scientific
19     Grocery & Gourmet Food
20               Pet Supplies

We have 21 categories in this dataset which are :

- Arts, Crafts & Sewing
- Cell Phones & Accessories
- Clothing, Shoes & Jewelry
- Tools & Home Improvement
- Health & Personal Care
- Baby Products
- Baby
- Patio, Lawn & Garden
- Beauty
- Sports & Outdoors
- Electronics
- All Electronics
- Automotive
- Toys & Games
- All Beauty
- Office Products
- Appliances
- Musical Instruments
- Industrial & Scientific
- Grocery & Gourmet Food
- Pet Supplies

In [13]:
products_data['categories'].value_counts()

Arts, Crafts & Sewing        2225
Beauty                       2202
Grocery & Gourmet Food       2201
Sports & Outdoors            2201
Automotive                   2200
Industrial & Scientific      2200
Musical Instruments          2200
Appliances                   2200
Office Products              2200
All Beauty                   2200
Toys & Games                 2200
Electronics                  2200
All Electronics              2200
Cell Phones & Accessories    2200
Patio, Lawn & Garden         2200
Baby                         2200
Baby Products                2200
Health & Personal Care       2200
Tools & Home Improvement     2200
Clothing, Shoes & Jewelry    2200
Pet Supplies                 2200
Name: categories, dtype: int64

We observes that our data is balanced because we have almost all categories with the same number of products(approximately 2200 products).

# Data cleaning 

In [14]:
products_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46229 entries, 0 to 46228
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         46228 non-null  object
 1   description  45187 non-null  object
 2   categories   46229 non-null  object
dtypes: object(3)
memory usage: 1.1+ MB


In [15]:
# check missing values for each column 
products_data.isnull().sum().sort_values(ascending=False)

description    1042
name              1
categories        0
dtype: int64

#### We observe that there is missing data in columns, we have 1042 product description and one product name are missing

In [16]:
# check out the rows with missing values
products_data[products_data.isnull().any(axis=1)].head()

name description  \
68                            Accentus: Transcriptions         NaN   
233  Nurtured by Nature Environmentally-Sensitive D...         NaN   
250             HP Photosmart C5580 All-in-One Printer         NaN   
257     Bounty Hunter QD2 Quick Draw II Metal Detector         NaN   
269        Shure WH20XLR Dynamic Microphone - Cardioid         NaN   

               categories  
68                 Beauty  
233         Baby Products  
250       All Electronics  
257  Patio, Lawn & Garden  
269   Musical Instruments

### Delet the rows that contain missing value

In [17]:
products_data = products_data.dropna()
products_data

name  \
0                       TUNGSTEN SOLDER PICK WITH HANDLE   
1      Write Right 98167 Screen Protector for Sony T615C   
2      Casio Mens DBC310-1 Databank 300 Digital Watch...   
3      Factory-Reconditioned DEWALT DW260KR Heavy-Dut...   
4                                 Energizer 2 in 1 Light   
...                                                  ...   
46224  Berkley Trilene XL 1000 - yard Fishing Line, C...   
46225       Badger Air-Brush Company Basic Spray Gun Set   
46226  Fimo Soft Polymer Clay 2 Ounces-8020-53 Tropic...   
46227               Wilton Icing Colors 1 Oz: Rose Petal   
46228  Polyform Sculpey III Polymer Clay 2 Oz: French...   

                                             description  \
0      Solder Pick for picking up molten solder when ...   
1      We all screen. And we all need to protect thos...   
2      Bringing you precision at a glance, the Casio ...   
3      Factory-Reconditioned DEWALT DW260KR Heavy-Dut...   
4      This twoway light features a bright flashlight...   
...                                                  ...   
46224  Smooth - casting Berkley Trilene XL Fishing Li...   
46225  An easy to use, top quality spray gun for the ...   
46226  Fimo Soft Polymer Clay is easier to use than b...   
46227  Icing Colors. This plastic bottle contains 1 o...   
46228  Sculpey 3 Polymer Clay is America's original o...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
46224          Sports & Outdoors  
46225      Arts, Crafts & Sewing  
46226      Arts, Crafts & Sewing  
46227     Grocery & Gourmet Food  
46228      Arts, Crafts & Sewing  

[45187 rows x 3 columns]

In [18]:
# check if missing values is deleted in rows and columns 
products_data.isnull().sum().sort_values(ascending=False)
products_data[products_data.isnull().any(axis=1)].head()

Empty DataFrame
Columns: [name, description, categories]
Index: []

In [19]:
products_data['categories'].value_counts()

Grocery & Gourmet Food       2196
Toys & Games                 2194
All Beauty                   2190
Cell Phones & Accessories    2188
Electronics                  2187
Tools & Home Improvement     2183
Clothing, Shoes & Jewelry    2182
Appliances                   2181
Sports & Outdoors            2178
Baby                         2177
All Electronics              2176
Arts, Crafts & Sewing        2175
Patio, Lawn & Garden         2168
Health & Personal Care       2154
Beauty                       2149
Office Products              2144
Baby Products                2131
Pet Supplies                 2124
Musical Instruments          2095
Industrial & Scientific      2016
Automotive                   1999
Name: categories, dtype: int64

In [20]:
products_data.shape

(45187, 3)

In [21]:
# Check for duplicated data
products_data.duplicated().sum()

224

In [22]:
# Delet duplicated data
products_data.drop_duplicates(keep=False, inplace = True)
products_data.duplicated().sum()

0

In [23]:
products_data.shape

(44804, 3)

Data still almost balanced, so we don't need to balance it

# Text preprocessing

### Lower casing, Tokenize, Stemming and Lemmatisation

In [24]:
import re
import nltk

lst_stopwords = nltk.corpus.stopwords.words("english")
lst_stopwords


def utils_preprocess_text(text, flg_stemm=False, flg_lemm=True, lst_stopwords=None):
    # lower casing
    text = re.sub(r'[^\w\s]', '', str(text).lower().strip())
            
    # Tokenize 
    lst_text = text.split()
    # remove Stopwords
    if lst_stopwords is not None:
        lst_text = [word for word in lst_text if word not in lst_stopwords]
                
    # Stemming 
    if flg_stemm == True:
        ps = nltk.stem.porter.PorterStemmer()
        lst_text = [ps.stem(word) for word in lst_text]
                
    # Lemmatisation 
    if flg_lemm == True:
        lem = nltk.stem.wordnet.WordNetLemmatizer()
        lst_text = [lem.lemmatize(word) for word in lst_text]
            
    # back to string from list
    text = " ".join(lst_text)
    return text

products_data["name"] = products_data["name"].apply(lambda x: utils_preprocess_text(x, flg_stemm=False, flg_lemm=True, lst_stopwords=lst_stopwords))
products_data["description"] = products_data["description"].apply(lambda x: utils_preprocess_text(x, flg_stemm=False, flg_lemm=True, lst_stopwords=lst_stopwords))
products_data

name  \
0                            tungsten solder pick handle   
1          write right 98167 screen protector sony t615c   
2      casio men dbc3101 databank 300 digital watch d...   
3      factoryreconditioned dewalt dw260kr heavyduty ...   
4                                    energizer 2 1 light   
...                                                  ...   
46224  berkley trilene xl 1000 yard fishing line clea...   
46225        badger airbrush company basic spray gun set   
46226  fimo soft polymer clay 2 ounces802053 tropical...   
46227                 wilton icing color 1 oz rose petal   
46228  polyform sculpey iii polymer clay 2 oz french ...   

                                             description  \
0       solder pick picking molten solder making jewelry   
1      screen need protect screen expensive little ga...   
2      bringing precision glance casio men databank d...   
3      factoryreconditioned dewalt dw260kr heavyduty ...   
4      twoway light feature bright flashlight beam ad...   
...                                                  ...   
46224  smooth casting berkley trilene xl fishing line...   
46225  easy use top quality spray gun beginner spray ...   
46226  fimo soft polymer clay easier use making perfe...   
46227  icing color plastic bottle contains 1 ounce ic...   
46228  sculpey 3 polymer clay america original oven b...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
46224          Sports & Outdoors  
46225      Arts, Crafts & Sewing  
46226      Arts, Crafts & Sewing  
46227     Grocery & Gourmet Food  
46228      Arts, Crafts & Sewing  

[44804 rows x 3 columns]

### Remove punctuation 

In [25]:
import string
PUNCT_TO_REMOVE = string.punctuation

def remove_punctuation(text):
    return text.translate(str.maketrans('', '', PUNCT_TO_REMOVE))


products_data["name"] = products_data["name"].apply(lambda text: remove_punctuation(text))
products_data["description"] = products_data["description"].apply(lambda text: remove_punctuation(text))
products_data

name  \
0                            tungsten solder pick handle   
1          write right 98167 screen protector sony t615c   
2      casio men dbc3101 databank 300 digital watch d...   
3      factoryreconditioned dewalt dw260kr heavyduty ...   
4                                    energizer 2 1 light   
...                                                  ...   
46224  berkley trilene xl 1000 yard fishing line clea...   
46225        badger airbrush company basic spray gun set   
46226  fimo soft polymer clay 2 ounces802053 tropical...   
46227                 wilton icing color 1 oz rose petal   
46228  polyform sculpey iii polymer clay 2 oz french ...   

                                             description  \
0       solder pick picking molten solder making jewelry   
1      screen need protect screen expensive little ga...   
2      bringing precision glance casio men databank d...   
3      factoryreconditioned dewalt dw260kr heavyduty ...   
4      twoway light feature bright flashlight beam ad...   
...                                                  ...   
46224  smooth casting berkley trilene xl fishing line...   
46225  easy use top quality spray gun beginner spray ...   
46226  fimo soft polymer clay easier use making perfe...   
46227  icing color plastic bottle contains 1 ounce ic...   
46228  sculpey 3 polymer clay america original oven b...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
46224          Sports & Outdoors  
46225      Arts, Crafts & Sewing  
46226      Arts, Crafts & Sewing  
46227     Grocery & Gourmet Food  
46228      Arts, Crafts & Sewing  

[44804 rows x 3 columns]

### Removal of URLs

In [26]:
def remove_urls(text):
    url_pattern = re.compile(r'https?://\S+|www\.\S+')
    return url_pattern.sub(r'', text)

products_data['name']=products_data['name'].apply(remove_urls)
products_data['description']=products_data['description'].apply(remove_urls)


### Remove numbers from data

In [27]:
# delete numbers from data
products_data['name'] = products_data['name'].str.replace('\d+', '')
products_data['description'] = products_data['description'].str.replace('\d+', '')
products_data

name  \
0                            tungsten solder pick handle   
1                  write right  screen protector sony tc   
2      casio men dbc databank  digital watch disconti...   
3      factoryreconditioned dewalt dwkr heavyduty  am...   
4                                      energizer   light   
...                                                  ...   
46224    berkley trilene xl  yard fishing line clear  lb   
46225        badger airbrush company basic spray gun set   
46226      fimo soft polymer clay  ounces tropical green   
46227                  wilton icing color  oz rose petal   
46228  polyform sculpey iii polymer clay  oz french blue   

                                             description  \
0       solder pick picking molten solder making jewelry   
1      screen need protect screen expensive little ga...   
2      bringing precision glance casio men databank d...   
3      factoryreconditioned dewalt dwkr heavyduty  am...   
4      twoway light feature bright flashlight beam ad...   
...                                                  ...   
46224  smooth casting berkley trilene xl fishing line...   
46225  easy use top quality spray gun beginner spray ...   
46226  fimo soft polymer clay easier use making perfe...   
46227  icing color plastic bottle contains  ounce ici...   
46228  sculpey  polymer clay america original oven ba...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
46224          Sports & Outdoors  
46225      Arts, Crafts & Sewing  
46226      Arts, Crafts & Sewing  
46227     Grocery & Gourmet Food  
46228      Arts, Crafts & Sewing  

[44804 rows x 3 columns]

In [28]:
# save the preprocessed data
products_data.to_csv('preprocessed data.csv',index=False)

# Data visualisation

In [29]:
products_data = pd.read_csv('preprocessed data.csv')
products_data

name  \
0                            tungsten solder pick handle   
1                  write right  screen protector sony tc   
2      casio men dbc databank  digital watch disconti...   
3      factoryreconditioned dewalt dwkr heavyduty  am...   
4                                      energizer   light   
...                                                  ...   
44799    berkley trilene xl  yard fishing line clear  lb   
44800        badger airbrush company basic spray gun set   
44801      fimo soft polymer clay  ounces tropical green   
44802                  wilton icing color  oz rose petal   
44803  polyform sculpey iii polymer clay  oz french blue   

                                             description  \
0       solder pick picking molten solder making jewelry   
1      screen need protect screen expensive little ga...   
2      bringing precision glance casio men databank d...   
3      factoryreconditioned dewalt dwkr heavyduty  am...   
4      twoway light feature bright flashlight beam ad...   
...                                                  ...   
44799  smooth casting berkley trilene xl fishing line...   
44800  easy use top quality spray gun beginner spray ...   
44801  fimo soft polymer clay easier use making perfe...   
44802  icing color plastic bottle contains  ounce ici...   
44803  sculpey  polymer clay america original oven ba...   

                      categories  
0          Arts, Crafts & Sewing  
1      Cell Phones & Accessories  
2      Clothing, Shoes & Jewelry  
3       Tools & Home Improvement  
4         Health & Personal Care  
...                          ...  
44799          Sports & Outdoors  
44800      Arts, Crafts & Sewing  
44801      Arts, Crafts & Sewing  
44802     Grocery & Gourmet Food  
44803      Arts, Crafts & Sewing  

[44804 rows x 3 columns]